eng/scripts/inject_codesnippets.py

```diff
@@ -0,0 +1,139 @@
+import os
+import glob
+import re
+import json
+from xml.sax.saxutils import escape, unescape
+
+# run this from the root of the repository
+root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
+
+SNIPPET_BEGIN = r"\s*\/\/\s*BEGIN\:\s*(?P<id>[a-zA-Z0-9\.\#\-\_]*)\s*"
+SNIPPET_END = r"\s*\/\/\s*END\:\s*(?P<id>[a-zA-Z0-9\.\#\-\_]*)\s*"
+SNIPPET_CALL = r"(?P<leadingspace>.*)\{\@codesnippet(?P<snippetid>.*)\}"
+
+EXCLUSION_ARRAY = ["JavadocCodeSnippetCheck.java"]
+
+HTML_ESCAPE_TABLE = {
+    "&": "&amp;",
+    '"': "&quot;",
+    ">": "&gt;",
+    "<": "&lt;",
+    "@": "{@literal @}",
+    "{": "&#123;",
+    "}": "&#125;",
+    "(": "&#40;",
+    ")": "&#41;",
+    "/": "&#47;",
+    "\\": "&#92;",
+}
+
+
+class SnippetDict:
+    def __init__(self, start_dict={}):
+        self.snippet_dict = start_dict
+
+    def __nonzero__(self):
+        return self.snippet_dict
+
+    def __str__(self):
+        return json.dumps(self.snippet_dict)
+
+    def begin_snippet(self, key):
+        if key not in self.snippet_dict:
+            self.snippet_dict[key] = []
+        else:
+            raise "Duplicate snippet begin."
+
+    def process_line(self, line):
+        self.snippet_dict = {k: v + [line] for k, v in self.snippet_dict.items()}
+
+    def finalize_snippet(self, key):
+        return self.snippet_dict.pop(key)
+
+
+def get_snippets_from_file(file):
+    finished_snippets = {}
+    running_dict = SnippetDict()
+
+    with open(file, "r", encoding="utf-8") as source:
+        print(file)
+        for line in source.readlines():
+
+            begin = re.match(SNIPPET_BEGIN, line)
+            end = re.match(SNIPPET_END, line)
+
+            if begin:
+                id_beginning = begin.groupdict()["id"]
+                running_dict.begin_snippet(id_beginning)
+            elif end:
+                id_ending = end.groupdict()["id"]
+                ending = running_dict.finalize_snippet(id_ending)
+                finished_snippets[id_ending] = ending
+            elif running_dict:
+                running_dict.process_line(line)
+
+    return finished_snippets
+
+
+def check_exclusion(file_name, exclusion_array):
+    if not os.path.isdir(file_name):
+        name = os.path.basename(file_name)
+
+        return name in exclusion_array
+
+
+if __name__ == "__main__":
+
+    # walk the codebase, find all java files
+    all_files = glob.glob("**/*.java", recursive=True)
```

Comment:
Right now the only directories that can contain snippets are in `src/samples/java/*` if we want to tighten where we are looking for snippets.

Reply:
You're right. Why I choose this method is that I need all the java files _anyway_ to search for @codesnippet CALLS.

Given that I need to find all the java files anyway, I do it once, then pull from that already generated list for the `snippets` files.

I will adjust the targeting in the filter though. Thanks!